### ISSUE TITLE
Question about Action Range and Scaling in AVG Implementation

### ISSUE THREAD

**Issue by SakodaShintaro** [gauthamvasan/avg#3]
Thank you for your great work and for publishing the code!
I am very interested in incremental reinforcement learning and your approach.

I have a question regarding the range of actions.
Since AVG uses a tanh squash policy, the range of actions is `[-1, 1]`.
On the other hand, when interacting with the environment, there seems to be no scaling factor applied.

https://github.com/gauthamvasan/avg/blob/main/avg.py

According to the Gymnasium Documentation, the actions space of `Humanoid` is defined as `Box(-0.4, 0.4)`, which seems inconsistent with the output.

For reference, other implementations, such as ClearnRL, scale the actions like this:
https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/sac_continuous_action.py#L113-L126

Does AVG not require such scaling?

In my experiments, applying scaling seems to worsen the learning performance.
However, I believe scaling might be desirable when applying the algorithm to other environments.

I'm sorry if I misunderstood.

Thank you!

**Comment by JagerHoHo** [comment]
It seems the code is missing an affine transformation to rescale the sampled actions back to match the environment's action space. This can be effectively addressed by utilizing torch.distributions to ensure the actions align with the expected range. Adding this step would help maintain consistency between the policy's outputs and the environment's requirements.

```py
from torch.distributions import Normal, TransformedDistribution, TanhTransform, AffineTransform

#-----------------------
# During init
self.action_scale = (env.single_action_space.high - env.single_action_space.low) / 2.0
self.action_bias = (env.single_action_space.high + env.single_action_space.low) / 2.0
#-----------------------

dist = TransformedDistribution(
    Normal(mu, torch.exp(log_std)), #can be MultivariateNormal like in avg.py
    [
        TanhTransform(cache_size=1),  # Squash to [-1, 1]
        AffineTransform(loc=self.action_bias, scale=self.action_scale),  # Scale to [act_low, act_high]
    ]
)

action = dist.rsample()
log_prob = dist.log_prob(action)
```

**Comment by SakodaShintaro** [comment]
@JagerHoHo 
Thank you for your comment. 

I have implemented your suggestion.

https://github.com/gauthamvasan/avg/compare/main...SakodaShintaro:avg:fix/action_range

I've just started experimenting, so I'll share the results as they come in. 😃

**Comment by JagerHoHo** [comment]
@SakodaShintaro 
If you are experimenting with the `avg.py`, I suggest you to add back the ScaleTDError part from the paper.
As mentioned in Issue #2, `avg.py` is missing this part of logic.

**Comment by gauthamvasan** [comment]
Thank you, @SakodaShintaro and @JagerHoHo, for your comments!

Initially, I had assumed that Mujoco environments would automatically apply an affine transformation internally, but your observations raise an important point.

> According to the Gymnasium Documentation, the action space of Humanoid is defined as Box(-0.4, 0.4), which seems inconsistent with the output.

I revisited this today and confirmed your observation. Using the range [−1,1] from the squashed normal distribution appears to perform better than the Box(-0.4, 0.4) range specified originally. This opens up a few interesting considerations:

1. The originally specified action range in Mujoco might not be ideal and should be modified.
2. Without explicitly enforcing the action space range, comparing results across different methods fairly is challenging.
3. The use of an affine transformation may require different hyperparameters to achieve optimal performance.

**Comment by gauthamvasan** [comment]
@SakodaShintaro

> Does AVG not require such scaling?

Like all other methods, AVG does require appropriate scaling. For example, we scale the actuation commands in real robot experiments. I had mistakenly assumed that Mujoco would automatically handle the scaling of actuation commands resulting from neural network policy actions, but that doesn’t seem to be the case.

**Comment by JagerHoHo** [comment]
@gauthamvasan 

> 1. The originally specified action range in Mujoco might not be ideal and should be modified.

I think it might be a special lucky case where the [-1, 1] can fully cover the action space [-0.4, 0.4]. For example, if it was [-4, 4], [-1, 1] can never fully cover all possible actions.

> 2. Without explicitly enforcing the action space range, comparing results across different methods fairly is challenging.

As far as I know, most rl environments for benchmarking will have a specified limited action range.

🤔

**Comment by gauthamvasan** [comment]
@JagerHoHo

> I think it might be a special lucky case where the [-1, 1] can fully cover the action space [-0.4, 0.4].

Yep! It was definitely lucky that all our other envs (except Humanoid-v4) had a range of [-1, 1].

> As far as I know, most RL environments for benchmarking will have a specified limited action range.

Right, let me explain what I meant. Take Reacher-v4 for example—if you send an action like [-1.5, 3.], the Mujoco base environment will just apply it directly without clamping it to [-1, 1]. I’m not talking about affine transforms here, just a simple clip/clamp operation to enforce the max/min actuation limits.

In some cases, like Humanoid, having these “larger” action spaces seems to help, even if the recommended range says otherwise. My suggestion was more about having Mujoco enforce this clamping directly.

As for affine transforms to scale actions properly, the agent implementation should take care of that.

I’ll add your suggested fix in the new PR along with the updates for issues #2 and #3.

**Comment by SakodaShintaro** [comment]
Thank you, guys.
I only applied the modification of action ranges. (without https://github.com/gauthamvasan/avg/issues/2)

Unfortunately, nan values emerged, and the experiment crashed.

```
E: 9746| D: 1.642| S: 235| R: 1538.44| T: 2044320
E: 9747| D: 0.906| S: 129| R: 770.82| T: 2044449
E: 9748| D: 2.121| S: 303| R: 1991.37| T: 2044752
E: 9749| D: 1.376| S: 196| R: 1182.29| T: 2044948
E: 9750| D: 2.335| S: 333| R: 2164.01| T: 2045281
E: 9751| D: 2.533| S: 363| R: 2362.27| T: 2045644
E: 9752| D: 6.985| S: 1000| R: 6730.97| T: 2046644
Expected parameter loc (Tensor of shape (1, 17)) of distribution MultivariateNormal(loc: torch.Size([1, 17]), covariance_matrix: torch.Size([1, 17, 17])) to satisfy the constraint IndependentConstraint(Real(), 1), but found invalid values:
tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]],
       grad_fn=<ExpandBackward0>)
Exiting this run, storing partial logs in the database for future debugging...
Appending partial episode #9752, length: 355, Total Steps: 2047001
Run with id: 20250128_100647-AVG-Humanoid-v4_seed-42 took 14352.767s!
```

Up to the crash step, the obtained return looks good.

![Image](https://github.com/user-attachments/assets/bd4e582c-4817-4439-96fc-5ed64f82e7ca)

In my previous experiments, I also implemented some other modifications regarding reward scaling (I used symlog squashing for rewards, inspired by [DreamerV3](https://vitalab.github.io/article/2023/01/19/DreamerV3.html)). This modification might not have worked well with the action range adjustment. I understand that due to entropy regularization, the hyperparameters are sensitive to the magnitude of the reward.

I will wait for @gauthamvasan 's update. I'm looking forward to it!

**Comment by gauthamvasan** [comment]
Addressed in #4. Thank you @SakodaShintaro  and @JagerHoHo !
